internal/xds/rbac/rbac_engine.go

```diff
@@ -19,24 +19,101 @@
 package rbac
 
 import (
+	"context"
+	"crypto/x509"
+	"errors"
 	"net"
+	"strconv"
 
 	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/metadata"
 	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
 )
 
-// Engine is used for matching incoming RPCs to policies.
-type Engine struct {
+// ChainedRBACEngine represents a chain of RBAC Engines, which will be used in
+// order to determine whether to allow incoming RPC's to proceed according to
+// the actions of each engine.
+type ChainedRBACEngine struct {
+	chainedEngines []*engine
+}
+
+// NewChainEngine returns a new Chain of RBAC Engines to query if incoming RPC's
+// are allowed to proceed or not. This function returns an error in the case
+// that one of the policies is invalid.
+func NewChainEngine(policies []*v3rbacpb.RBAC) (*ChainedRBACEngine, error) {
+	var chainedEngines []*engine
+	for _, policy := range policies {
+		rbacEngine, err := newEngine(policy)
+		if err != nil {
+			return nil, err
+		}
+		chainedEngines = append(chainedEngines, rbacEngine)
+	}
+	return &ChainedRBACEngine{chainedEngines: chainedEngines}, nil
+}
+
+// DetermineStatus takes in data about incoming RPC's and returns a status error
+// representing whether the RPC should be denied or not (i.e. PermissionDenied or OK)
+// based on the full list of RBAC Engines and their associated actions.
+func (cre *ChainedRBACEngine) DetermineStatus(data Data) error {
+	// This conversion step (i.e. pulling things out of generic ctx) can be done
+	// once, and then be used for the whole chain of RBAC Engines.
+	rpcData, err := newRPCData(data)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, "passed in data did not have enough information representing incoming rpc")
+	}
+	for _, engine := range cre.chainedEngines {
+		// TODO: What do I do now with this matchingPolicyName?
```

Comment:
So the thing is, the matchingPolicyName is per engine. If you were to return matchingPolicyNames, it would be a slice of matchingPolicyNames from here. I find that not very useful to the interceptor. Thus, I don't know what to do with them here.

Reply:
Great idea :)! Added return to error message about matching DENY policy.